### ISSUE TITLE
all pooled connections were in use

### ISSUE THREAD

**Issue by sandroshu** [postalserver/postal#2631]
## Describe the bug

Timeout error while processing unqueue

## To Reproduce

1. Send mail
2. Open worker log
3. See error

## Expected behaviour

Process the queue

## Environment details 

Docker with 1 worker

## Additional information/context

`[9] [2023-10-02T08:26:55.397] INFO -- : [BLQPEM9OOP] Started processing UnqueueMessageJob job
[9] [2023-10-02T08:26:55.391] WARN -- : [TF1EL6OITL] ActiveRecord::ConnectionTimeoutError: could not obtain a connection from the pool within 5.000 seconds (waited 5.001 seconds); all pooled connections were in use
[9] [2023-10-02T08:26:55.398] WARN -- : [TF1EL6OITL]    /usr/local/bundle/gems/activerecord-6.1.7.3/lib/active_record/connection_adapters/abstract/connection_pool.rb:210:in `block in wait_poll'
[9] [2023-10-02T08:26:55.398] WARN -- : [TF1EL6OITL]    /usr/local/bundle/gems/activerecord-6.1.7.3/lib/active_record/connection_adapters/abstract/connection_pool.rb:199:in `loop'
[9] [2023-10-02T08:26:55.398] WARN -- : [TF1EL6OITL]    /usr/local/bundle/gems/activerecord-6.1.7.3/lib/active_record/connection_adapters/abstract/connection_pool.rb:199:in `wait_poll'
[9] [2023-10-02T08:26:55.398] WARN -- : [TF1EL6OITL]    /usr/local/bundle/gems/activerecord-6.1.7.3/lib/active_record/connection_adapters/abstract/connection_pool.rb:160:in `internal_poll'
[9] [2023-10-02T08:26:55.399] WARN -- : [TF1EL6OITL]    /usr/local/bundle/gems/activerecord-6.1.7.3/lib/active_record/connection_adapters/abstract/connection_pool.rb:286:in `internal_poll'
[9] [2023-10-02T08:26:55.399] WARN -- : [TF1EL6OITL]    /usr/local/bundle/gems/activerecord-6.1.7.3/lib/active_record/connection_adapters/abstract/connection_pool.rb:155:in `block in poll'
[9] [2023-10-02T08:26:55.399] WARN -- : [TF1EL6OITL]    /usr/local/lib/ruby/3.2.0/monitor.rb:202:in `synchronize'
[9] [2023-10-02T08:26:55.399] WARN -- : [TF1EL6OITL]    /usr/local/lib/ruby/3.2.0/monitor.rb:202:in `mon_synchronize'
[9] [2023-10-02T08:26:55.399] WARN -- : [TF1EL6OITL]    /usr/local/bundle/gems/activerecord-6.1.7.3/lib/active_record/connection_adapters/abstract/connection_pool.rb:164:in `synchronize'
[9] [2023-10-02T08:26:55.399] WARN -- : [TF1EL6OITL]    /usr/local/bundle/gems/activerecord-6.1.7.3/lib/active_record/connection_adapters/abstract/connection_pool.rb:155:in `poll'
[9] [2023-10-02T08:26:55.399] WARN -- : [TF1EL6OITL]    /usr/local/bundle/gems/activerecord-6.1.7.3/lib/active_record/connection_adapters/abstract/connection_pool.rb:870:in `acquire_connection'
[9] [2023-10-02T08:26:55.400] WARN -- : [TF1EL6OITL]    /usr/local/bundle/gems/activerecord-6.1.7.3/lib/active_record/connection_adapters/abstract/connection_pool.rb:588:in `checkout'
[9] [2023-10-02T08:26:55.400] WARN -- : [TF1EL6OITL]    /usr/local/bundle/gems/activerecord-6.1.7.3/lib/active_record/connection_adapters/abstract/connection_pool.rb:428:in `connection'
[9] [2023-10-02T08:26:55.400] WARN -- : [TF1EL6OITL]    /usr/local/bundle/gems/activerecord-6.1.7.3/lib/active_record/connection_adapters/abstract/connection_pool.rb:1128:in `retrieve_connection'
[9] [2023-10-02T08:26:55.400] WARN -- : [TF1EL6OITL]    /usr/local/bundle/gems/activerecord-6.1.7.3/lib/active_record/connection_handling.rb:327:in `retrieve_connection'
[9] [2023-10-02T08:26:55.400] WARN -- : [TF1EL6OITL]    /usr/local/bundle/gems/activerecord-6.1.7.3/lib/active_record/connection_handling.rb:283:in `connection'
[9] [2023-10-02T08:26:55.400] WARN -- : [TF1EL6OITL]    /usr/local/bundle/gems/activerecord-6.1.7.3/lib/active_record/core.rb:490:in `cached_find_by_statement'
[9] [2023-10-02T08:26:55.401] WARN -- : [TF1EL6OITL]    /usr/local/bundle/gems/activerecord-6.1.7.3/lib/active_record/core.rb:386:in `find_by'
[9] [2023-10-02T08:26:55.401] WARN -- : [TF1EL6OITL]    /usr/local/bundle/gems/activerecord-6.1.7.3/lib/active_record/dynamic_matchers.rb:66:in `find_by_id'
[9] [2023-10-02T08:26:55.401] WARN -- : [TF1EL6OITL]    /opt/postal/app/app/jobs/unqueue_message_job.rb:4:in `perform'
[9] [2023-10-02T08:26:55.401] WARN -- : [TF1EL6OITL]    /opt/postal/app/lib/postal/worker.rb:64:in `receive_job'
[9] [2023-10-02T08:26:55.401] WARN -- : [TF1EL6OITL]    /opt/postal/app/lib/postal/worker.rb:96:in `block in join_queue'
[9] [2023-10-02T08:26:55.401] WARN -- : [TF1EL6OITL]    /usr/local/bundle/gems/bunny-2.20.3/lib/bunny/consumer.rb:56:in `call'
[9] [2023-10-02T08:26:55.402] WARN -- : [TF1EL6OITL]    /usr/local/bundle/gems/bunny-2.20.3/lib/bunny/channel.rb:1842:in `block in handle_frameset'
[9] [2023-10-02T08:26:55.402] WARN -- : [TF1EL6OITL]    /usr/local/bundle/gems/bunny-2.20.3/lib/bunny/consumer_work_pool.rb:108:in `block (2 levels) in run_loop'
[9] [2023-10-02T08:26:55.402] WARN -- : [TF1EL6OITL]    /usr/local/bundle/gems/bunny-2.20.3/lib/bunny/consumer_work_pool.rb:103:in `loop'
[9] [2023-10-02T08:26:55.402] WARN -- : [TF1EL6OITL]    /usr/local/bundle/gems/bunny-2.20.3/lib/bunny/consumer_work_pool.rb:103:in `block in run_loop'
[9] [2023-10-02T08:26:55.403] WARN -- : [TF1EL6OITL]    /usr/local/bundle/gems/bunny-2.20.3/lib/bunny/consumer_work_pool.rb:102:in `catch'
[9] [2023-10-02T08:26:55.403] WARN -- : [TF1EL6OITL]    /usr/local/bundle/gems/bunny-2.20.3/lib/bunny/consumer_work_pool.rb:102:in `run_loop'
[9] [2023-10-02T08:26:55.403] INFO -- : [TF1EL6OITL] Finished processing UnqueueMessageJob job in 5.013552212s`

**Comment by hugoperes** [comment]
I'm having the same problem. Is it a general problem with the Postal Server?

**Comment by hugoperes** [comment]
Guys, problem solved. It was a problem with my hosting and not with the Postal Server.

**Comment by willpower232** [comment]
@hugoperes are you able to elaborate on what the issue was in case it is relevant here?

**Comment by hugoperes** [comment]
@willpower232 Then the problem repeated itself again. A timeout simply occurs when trying to verify the postal server connection using a user credential. At first, I believe it is due to the hosting, but as was repeated today, there is a possibility that it is a problem with the postal server.

**Comment by sandroshu** [comment]
Currently all mails are pending and this is the last line in the workers log (restarting workers make it work for some time):

`#<Bunny::Session:0xca58 postal@127.0.0.1:5672, vhost=postal, addresses=[127.0.0.1:5672]>: Uncaught exception from consumer #<Bunny::Consumer:778060 @channel_id=1 @queue=deliver-jobs-outgoing-6 @consumer_tag=bunny-1696328771000-460158659925>: #<Bunny::ChannelAlreadyClosed: cannot use a closed channel! Channel id: 1, closed due to a server-reported channel error: PRECONDITION_FAILED - delivery acknowledgement on channel 1 timed out. Timeout value used: 1800000 ms. This timeout value can be configured, see consumers doc guide to learn more> @ /usr/local/bundle/gems/bunny-2.20.3/lib/bunny/channel.rb:2075:in raise_if_no_longer_open!'
[9] [2023-10-03T17:32:56.950] INFO -- : [ZNNWRFIO] Net::WriteTimeout: Net::WriteTimeout with #<TCPSocket:fd 17, AF_INET, x.x.x.x, 42741>
[9] [2023-10-03T17:33:05.318] INFO -- : [OACMZNJV3P] [16::41106 1991297] Message requeued for trying later.
[9] [2023-10-03T17:33:05.329] INFO -- : [ZNNWRFIO] Finishing up
[9] [2023-10-03T17:33:05.406] INFO -- : [OACMZNJV3P] Finished processing UnqueueMessageJob job in 1988.416179855s
E, [2023-10-03T17:33:05.406787 #9] ERROR -- #<Bunny::Session:0xca58 postal@127.0.0.1:5672, vhost=postal, addresses=[127.0.0.1:5672]>: Uncaught exception from consumer #<Bunny::Consumer:778060 @channel_id=1 @queue=deliver-jobs-outgoing-6 @consumer_tag=bunny-1696328771000-460158659925>: #<Bunny::ChannelAlreadyClosed: cannot use a closed channel! Channel id: 1, closed due to a server-reported channel error: PRECONDITION_FAILED - delivery acknowledgement on channel 1 timed out. Timeout value used: 1800000 ms. This timeout value can be configured, see consumers doc guide to learn more> @ /usr/local/bundle/gems/bunny-2.20.3/lib/bunny/channel.rb:2075:in `raise_if_no_longer_open!'`

**Comment by willpower232** [comment]
@hugoperes let us know if the problem re-occurs, it would be interesting to know whether there is an issue with the size of the table or whether there could have been slightly corrupt data in one of the rows which was making using the database server difficult

**Comment by hugoperes** [comment]
@willpower232 So, the problem was still persisting. One thing I was noticing was that the SMTP connection check was having a delay. Sometimes it even gave a timeout.

So I did something here that apparently solved it:

I edited docker-compose, placing the restart: unless-stopped parameter in the worker container.

I restarted the postcard with 8 workers.

When testing the SMTP connection now, it returns a 200 response very quickly. It's not giving a timeout.

I'm suspecting that the queued email history was crashing the workers and they weren't automatically restarting.
